### PR TITLE
Fix race condition on ssh key delete, and revert re-revert of hard line break

### PR DESCRIPTION
**Pull request by donbowman**
https://github.com/gogits/gogs/commit/c5be40decbfdb488b79bb4f53bf9d238fae20f0c re-added the hard line breaks to the markdown, this is undesirable (it was just fixed in previous pull request).

Also, there is a race condition in the ssh keys... on deleting one key, between writing the temp file and the rename, a new key can be added, and this will delete it.

I don't believe this fix is entirely correct, there is a problem still. one problem is that a user can have 2 keys w/ the same fingerprint, and the system will pick one (e.g. key-####) and then lookup (key-YYYY) and get an error that they don't own it.

I think the correct fix is to always rewrite the whole file from db (which code exists but is currently manual). But i didn't put that in the pull.

### CODE REVIEW

File: models/publickey.go [hunk 1]
```diff
@@ -386,8 +386,6 @@ func ListPublicKeys(uid int64) ([]*PublicKey, error) {
 
 // rewriteAuthorizedKeys finds and deletes corresponding line in authorized_keys file.
 func rewriteAuthorizedKeys(key *PublicKey, p, tmpP string) error {
-	sshOpLocker.Lock()
-	defer sshOpLocker.Unlock()
```
Comment (Unknwon) [inline models/publickey.go]: This empty line should be removed.

**Comment by Unknwon** [comment]
Thanks your PR!

I've made some comments to help your PR be better.

Besides,

These [two lines](https://github.com/donbowman/gogs/blob/master/models/publickey.go#L267) should be removed, and add them to two other places:

- https://github.com/donbowman/gogs/blob/master/models/publickey.go#L337
- https://github.com/donbowman/gogs/blob/master/models/publickey.go#L588

**Comment by donbowman** [comment]
i'm not sure i understand. the lock is used to control concurrent access to ~/.ssh/authorized_keys, why would the lock move to those spots?

**Comment by Unknwon** [comment]
OK, I see. My bad :flushed:

Just remove that empty line should be fine.

**Comment by Unknwon** [comment]
ping

**Comment by donbowman** [comment]
blank line removed as per req

**Comment by Unknwon** [comment]
Thank you very much!

**Comment by manfer** [comment]
```
	extensions |= blackfriday.EXTENSION_HARD_LINE_BREAK
```
Had been removed with this merge.

There is no problem to have that extension enabled or disabled. People will have to adapt their markdown to one situation or the other. But blackfriday with hard line breaks ON renders markdown more similar to what people are used to which is github flavoured markdown.

It is ok to remove it but given a reason would be the very least. This is undesirable doesn't look to me a  good reason. Undesirable for whom or for what reason exactly? This was fixed in previous PR is not accurate. This is not a fix of anything because it is not a bug having it turned on. At most it would be an enhancement if it is considered that turning hard line break off is an enhancement.

Maybe the best solution, if possible, is to add an option to the [config](http://gogs.io/docs/advanced/configuration_cheat_sheet.html) or in the admin settings so anyway can choose to enable or disable hard line breaks at will.

**Comment by tboerger** [comment]
I'm still searching for a reason why it have been removed

**Comment by Unknwon** [comment]
I would really like to add a repo-level settings as someone else has proposed on another thread.

**Comment by manfer** [comment]
> I would really like to add a repo-level settings as someone else has proposed on another thread.

That's even more fine grained option though I would include a way to config the default value in global config or admin settings anyway.

**Comment by Unknwon** [comment]
@manfer per user maybe make more sense since different users have different tastes.

**Comment by Unknwon** [comment]
@manfer but yes, global config is easier to accomplish, would be my final choice.

**Comment by manfer** [comment]
> @manfer per user maybe make more sense since different users have different tastes.

They are not mutually exclusive. First the admin, the owner of gogs installation, sets the default value, ON or OFF. Then users have the option to decide to set it ON or OFF at will for themselves.

The first should be very easy to implement. Just an if based on the setting should be enough.

The user option is a larger task and can be implemented later if needed.

**Comment by Unknwon** [comment]
OK, let move on to https://github.com/gogits/gogs/issues/1009

**Pull request merged**